File: apps/nexkeylock-pwa/src/lib/synchro.ts
/**
 * Client de **synchronisation zéro-connaissance** de la PWA.
 *
 * Appels en **même origine** vers `/sync/...` (proxy en dev, reverse-proxy en
 * prod) afin de garder la CSP stricte (`connect-src 'self'`). Le hash
 * d'authentification est dérivé **dans le WASM** (le mot de passe ne quitte pas
 * l'appareil) ; le blob poussé est le coffre **déjà chiffré**.
 */
import { coeur } from "./coeur";

const RACINE = "/sync";
const CLE_REVISION = "nexkeylock.sync.revision";
const CLE_EMAIL = "nexkeylock.sync.email";

let jeton: string | null = null;

/** Révision locale connue (concurrence optimiste). */
export function revisionLocale(): number {
  return Number(localStorage.getItem(CLE_REVISION) ?? "0");
}
function definirRevision(r: number) {
  localStorage.setItem(CLE_REVISION, String(r));
}
/** Email mémorisé (préférence, non secret). */
export function emailMemorise(): string {
  return localStorage.getItem(CLE_EMAIL) ?? "";
}
/** Vrai si une session de synchronisation est active. */
export function connecte(): boolean {
  return jeton !== null;
}

function versHex(octets: Uint8Array): string {
  return Array.from(octets, (o) => o.toString(16).padStart(2, "0")).join("");
}
function depuisHex(hex: string): Uint8Array {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

async function poster(chemin: string, corps: unknown, avecJeton = false): Promise<Response> {
  const entetes: Record<string, string> = { "Content-Type": "application/json" };
  if (avecJeton && jeton) entetes.Authorization = `Bearer ${jeton}`;
  return fetch(`${RACINE}${chemin}`, {
    method: "POST",
    headers: entetes,
    body: JSON.stringify(corps),
  });
}

/** Inscrit un compte (email + mot de passe maître). */
export async function inscrire(email: string, motDePasse: string): Promise<void> {
  const hash = await coeur.hashAuth(email, motDePasse);
  const rep = await poster("/inscription", { email, hash_auth: hash });
  if (rep.status === 409) throw "Un compte existe déjà pour cet email.";
  if (!rep.ok) throw "Inscription refusée par le serveur.";
}

/** Se connecte ; mémorise le jeton de session et l'email. */
export async function connecter(email: string, motDePasse: string): Promise<void> {
  const hash = await coeur.hashAuth(email, motDePasse);
  const rep = await poster("/connexion", { email, hash_auth: hash });
  if (!rep.ok) throw "Identifiants de synchronisation invalides.";
  const v = (await rep.json()) as { jeton?: string };
  if (!v.jeton) throw "Jeton absent de la réponse.";
  jeton = v.jeton;
  localStorage.setItem(CLE_EMAIL, email);
}

/** Tire le coffre distant ; renvoie (révision, octets) — octets vide si rien. */
export async function tirer(): Promise<{ revision: number; octets: Uint8Array }> {
  const rep = await fetch(`${RACINE}/coffre`, { headers: { Authorization: `Bearer ${jeton}` } });
  if (!rep.ok) throw "Session expirée ; reconnectez-vous.";
  const v = (await rep.json()) as { revision: number; blob: string };
  definirRevision(v.revision);
  return { revision: v.revision, octets: v.blob ? depuisHex(v.blob) : new Uint8Array() };
}

/** Pousse `octets` (coffre chiffré). `accepte=false` => conflit (révision distante). */
export async function pousser(octets: Uint8Array): Promise<{ accepte: boolean; revision: number }> {
  const rep = await poster("/coffre", { base: revisionLocale(), blob: versHex(octets) }, true);
  if (rep.status === 401) throw "Session expirée ; reconnectez-vous.";
  const v = (await rep.json()) as { revision?: number; actuelle?: number };
  if (rep.ok) {
    const revision = v.revision ?? 0;
    definirRevision(revision);
    return { accepte: true, revision };
  }
  // 409 : conflit.
  return { accepte: false, revision: v.actuelle ?? 0 };
}

/** Force l'envoi en écrasant le distant (résolution de conflit « garder local »). */
export async function forcer(octets: Uint8Array): Promise<{ accepte: boolean; revision: number }> {
  // S'aligne sur la révision distante courante, puis pousse.
  const distant = await tirer();
  definirRevision(distant.revision);
  return pousser(octets);
}

File: apps/nexkeylock-bureau/src-tauri/src/synchro.rs
//! Client de **synchronisation zéro-connaissance** (vers `nex-serveur-sync`).
//!
//! Le mot de passe maître ne quitte **jamais** l'appareil : on en dérive (via
//! Argon2id, sel issu de l'email, puis HKDF) un **hash d'authentification**
//! envoyé au serveur, indépendant de toute clé de chiffrement. Le « blob » poussé
//! est le **coffre déjà chiffré** (opaque pour le serveur).

use nex_cryptographie::kdf::{deriver_cle, ParametresArgon2};
use nex_sync::hash_authentification;
use sha2::{Digest, Sha256};

use crate::erreur::ErreurCommande;

/// Résultat d'un envoi (push).
pub enum ResultatPoussee {
    /// Accepté ; nouvelle révision.
    Accepte(u64),
    /// Conflit ; révision actuelle du serveur (tirer puis réessayer/forcer).
    Conflit(u64),
}

/// Paramètres Argon2id (allégés en mode test, comme ailleurs dans l'app).
fn parametres() -> ParametresArgon2 {
    if std::env::var_os("NEXKEYLOCK_KDF_RAPIDE").is_some() {
        ParametresArgon2::new(8, 1, 1)
    } else {
        ParametresArgon2::default()
    }
}

/// Sel déterministe dérivé de l'email (≥ 16 octets) : identique sur tous les
/// appareils d'un même compte, jamais secret.
fn sel_email(email: &str) -> Vec<u8> {
    let mut h = Sha256::new();
    h.update(b"nexkeylock-sync-sel:v1");
    h.update(email.trim().to_lowercase().as_bytes());
    h.finalize()[..16].to_vec()
}

/// Dérive le hash d'authentification (hex) à partir de l'email et du mot de passe.
fn hash_auth_hex(email: &str, mot_de_passe: &str) -> Result<String, ErreurCommande> {
    let cle_maitre = deriver_cle(mot_de_passe.as_bytes(), &sel_email(email), parametres())
        .map_err(|_| ErreurCommande::interne("Dérivation impossible."))?;
    let hash = hash_authentification(&cle_maitre)
        .map_err(|_| ErreurCommande::interne("Dérivation d'authentification impossible."))?;
    Ok(hex::encode(hash.exposer()))
}

fn reseau(_e: impl std::fmt::Debug) -> ErreurCommande {
    ErreurCommande::interne("Serveur de synchronisation injoignable.")
}

/// Analyse le corps JSON d'une réponse (minreq sans la feature json).
fn parse_corps(rep: &minreq::Response) -> Result<serde_json::Value, ErreurCommande> {
    let texte = rep
        .as_str()
        .map_err(|_| ErreurCommande::interne("Réponse du serveur illisible."))?;
    serde_json::from_str(texte)
        .map_err(|_| ErreurCommande::interne("Réponse du serveur illisible."))
}

/// Inscrit un compte sur le serveur.
pub fn inscrire(base: &str, email: &str, mot_de_passe: &str) -> Result<(), ErreurCommande> {
    let hash = hash_auth_hex(email, mot_de_passe)?;
    let corps = serde_json::json!({ "email": email, "hash_auth": hash }).to_string();
    let rep = minreq::post(format!("{base}/inscription"))
        .with_header("Content-Type", "application/json")
        .with_body(corps)
        .with_timeout(15)
        .send()
        .map_err(reseau)?;
    match rep.status_code {
        201 => Ok(()),
        409 => Err(ErreurCommande::interne(
            "Un compte existe déjà pour cet email.",
        )),
        _ => Err(ErreurCommande::interne(
            "Inscription refusée par le serveur.",
        )),
    }
}

/// Se connecte et renvoie un jeton de session.
pub fn connecter(base: &str, email: &str, mot_de_passe: &str) -> Result<String, ErreurCommande> {
    let hash = hash_auth_hex(email, mot_de_passe)?;
    let corps = serde_json::json!({ "email": email, "hash_auth": hash }).to_string();
    let rep = minreq::post(format!("{base}/connexion"))
        .with_header("Content-Type", "application/json")
        .with_body(corps)
        .with_timeout(15)
        .send()
        .map_err(reseau)?;
    if rep.status_code != 200 {
        return Err(ErreurCommande::interne(
            "Identifiants de synchronisation invalides.",
        ));
    }
    let v = parse_corps(&rep)?;
    v.get("jeton")
        .and_then(|j| j.as_str())
        .map(str::to_string)
        .ok_or_else(|| ErreurCommande::interne("Jeton absent de la réponse."))
}

/// Récupère (révision, blob chiffré) du serveur ; blob vide si rien à distance.
pub fn tirer(base: &str, jeton: &str) -> Result<(u64, Vec<u8>), ErreurCommande> {
    let rep = minreq::get(format!("{base}/coffre"))
        .with_header("Authorization", format!("Bearer {jeton}"))
        .with_timeout(30)
        .send()
        .map_err(reseau)?;
    if rep.status_code != 200 {
        return Err(ErreurCommande::interne(
            "Session expirée ; reconnectez-vous.",
        ));
    }
    let v = parse_corps(&rep)?;
    let revision = v.get("revision").and_then(|r| r.as_u64()).unwrap_or(0);
    let blob_hex = v.get("blob").and_then(|b| b.as_str()).unwrap_or("");
    let blob = if blob_hex.is_empty() {
        Vec::new()
    } else {
        hex::decode(blob_hex).map_err(|_| ErreurCommande::interne("Blob distant invalide."))?
    };
    Ok((revision, blob))
}

/// Pousse `blob` (coffre chiffré) en supposant la révision de base `base_rev`.
pub fn pousser(
    base: &str,
    jeton: &str,
    base_rev: u64,
    blob: &[u8],
) -> Result<ResultatPoussee, ErreurCommande> {
    let corps = serde_json::json!({ "base": base_rev, "blob": hex::encode(blob) }).to_string();
    let rep = minreq::post(format!("{base}/coffre"))
        .with_header("Authorization", format!("Bearer {jeton}"))
        .with_header("Content-Type", "application/json")
        .with_body(corps)
        .with_timeout(30)
        .send()
        .map_err(reseau)?;
    match rep.status_code {
        200 => {
            let revision = parse_corps(&rep)?
                .get("revision")
                .and_then(|r| r.as_u64())
                .unwrap_or(0);
            Ok(ResultatPoussee::Accepte(revision))
        }
        409 => {
            let actuelle = parse_corps(&rep)?
                .get("actuelle")
                .and_then(|r| r.as_u64())
                .unwrap_or(0);
            Ok(ResultatPoussee::Conflit(actuelle))
        }
        401 => Err(ErreurCommande::interne(
            "Session expirée ; reconnectez-vous.",
        )),
        _ => Err(ErreurCommande::interne("Envoi refusé par le serveur.")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hash_auth_deterministe_et_sensible() {
        std::env::set_var("NEXKEYLOCK_KDF_RAPIDE", "1");
        let a = hash_auth_hex("moi@ex.fr", "motdepasse").unwrap();
        let b = hash_auth_hex("moi@ex.fr", "motdepasse").unwrap();
        assert_eq!(a, b, "déterministe pour le même couple");
        // L'email (casse/espaces) est normalisé.
        let c = hash_auth_hex("  MOI@EX.FR ", "motdepasse").unwrap();
        assert_eq!(a, c);
        // Mot de passe ou email différents => hash différent.
        assert_ne!(a, hash_auth_hex("moi@ex.fr", "autre").unwrap());
        assert_ne!(a, hash_auth_hex("autre@ex.fr", "motdepasse").unwrap());
        // 32 octets => 64 hex.
        assert_eq!(a.len(), 64);
    }

    #[test]
    fn synchro_complete_contre_un_vrai_serveur() {
        std::env::set_var("NEXKEYLOCK_KDF_RAPIDE", "1");
        // Démarre un vrai serveur en mémoire sur un port éphémère (loopback).
        let serveur = nex_serveur_sync::lier("127.0.0.1:0").expect("liaison serveur");
        let adresse = serveur.server_addr().to_ip().expect("adresse ip");
        std::thread::spawn(move || nex_serveur_sync::servir(serveur));
        let base = format!("http://{adresse}");

        // Inscription + connexion.
        inscrire(&base, "moi@ex.fr", "motdepasse").unwrap();
        let jeton = connecter(&base, "moi@ex.fr", "motdepasse").unwrap();

        // Mauvais mot de passe rejeté.
        assert!(connecter(&base, "moi@ex.fr", "faux").is_err());

        // Premier envoi (base 0) accepté en révision 1.
        let blob = b"coffre-chiffre-opaque";
        match pousser(&base, &jeton, 0, blob).unwrap() {
            ResultatPoussee::Accepte(r) => assert_eq!(r, 1),
            ResultatPoussee::Conflit(_) => panic!("conflit inattendu"),
        }

        // Tirage : on récupère exactement le blob opaque envoyé.
        let (revision, recu) = tirer(&base, &jeton).unwrap();
        assert_eq!(revision, 1);
        assert_eq!(recu, blob);

        // Renvoi sur base périmée (0) => conflit, révision actuelle 1.
        match pousser(&base, &jeton, 0, blob).unwrap() {
            ResultatPoussee::Conflit(actuelle) => assert_eq!(actuelle, 1),
            ResultatPoussee::Accepte(_) => panic!("aurait dû être un conflit"),
        }
    }
}
